### PR TITLE
Support evaluating min/max only metadata query

### PR DESCRIPTION
**Pull request by shixuan-fan**
```
== RELEASE NOTES ==

General Changes
* Add support to optimize min/max only metadata query. This is controlled by existing config ``optimizer.optimize-metadata-queries`` and session property ``optimize_metadata_queries``. Note that enabling this config/session property might change query result if there are metadata that refers to empty data, e.g. empty hive partition.
```

### CODE REVIEW

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java [hunk 1]
```diff
@@ -277,6 +280,83 @@ public void testPartitionPruning()
...
+                .build();
+
+        queryRunner.execute(shufflePartitionColumns, "CREATE TABLE test_metadata_aggregation_folding WITH (partitioned_by = ARRAY['ds']) AS " +
+                "SELECT orderkey, CAST(to_iso8601(date_add('DAY', orderkey % 7, date('2020-07-01'))) AS VARCHAR) AS ds FROM orders WHERE orderkey < 1000");
+        queryRunner.execute(shufflePartitionColumns, "CREATE TABLE test_metadata_aggregation_folding_more_partitions WITH (partitioned_by = ARRAY['ds']) AS " +
+                "SELECT orderkey, CAST(to_iso8601(date_add('DAY', orderkey % 200, date('2020-07-01'))) AS VARCHAR) AS ds FROM orders WHERE orderkey < 1000");
+        queryRunner.execute(shufflePartitionColumns, "CREATE TABLE test_metadata_aggregation_folding_null_partitions WITH (partitioned_by = ARRAY['ds']) AS " +
+                "SELECT orderkey, CAST(to_iso8601(date_add('DAY', orderkey % 7, date('2020-07-01'))) AS VARCHAR) AS ds FROM orders WHERE orderkey < 1000");
+        queryRunner.execute(shufflePartitionColumns, "INSERT INTO test_metadata_aggregation_folding_null_partitions SELECT 0 as orderkey, null AS ds");
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java:300]: Breakdown the params into multiple lines
```java
        queryRunner.execute(
                shufflePartitionColumns,
                "CREATE TABLE test_metadata_aggregation_folding WITH (partitioned_by = ARRAY['ds']) AS " +
                        "SELECT orderkey, CAST(to_iso8601(date_add('DAY', orderkey % 7, date('2020-07-01'))) AS VARCHAR) AS ds FROM orders WHERE orderkey < 1000");

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java [hunk 2]
```diff
@@ -151,31 +166,111 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> cont
                 return context.defaultRewrite(node);
             }
 
-            ImmutableList.Builder<List<RowExpression>> rowsBuilder = ImmutableList.builder();
-            for (TupleDomain<ColumnHandle> domain : predicates.getPredicates()) {
-                if (!domain.isNone()) {
-                    Map<ColumnHandle, NullableValue> entries = TupleDomain.extractFixedValues(domain).get();
+            if (!isReducible(node)) {
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java:169]: Maybe add a helper function when `isReducible(node)` is true and put the new logic there? This function is already pretty big.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java [hunk 3]
```diff
@@ -151,31 +166,111 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> cont
...
+            }
+
+            // Fold min/max aggregations to a constant value
+            ImmutableList.Builder<RowExpression> scalarsBuilder = ImmutableList.builder();
+            for (int i = 0; i < inputs.size(); i++) {
+                ImmutableList.Builder<RowExpression> arguments = ImmutableList.builder();
+                ColumnHandle column = columns.get(inputs.get(i));
+                // for each input column, add a literal expression using the entry value
+                for (TupleDomain<ColumnHandle> domain : predicates.getPredicates()) {
+                    if (!domain.isNone()) {
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java:203]:
```java
if (domain.isNone()) {
    continue;
}

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java [hunk 4]
```diff
@@ -151,31 +166,111 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> cont
...
+            for (Aggregation aggregation : node.getAggregations().values()) {
+                FunctionMetadata functionMetadata = metadata.getFunctionManager().getFunctionMetadata(aggregation.getFunctionHandle());
+                if (!AGGREGATION_SCALAR_MAPPING.containsKey(functionMetadata.getName()) || functionMetadata.getArgumentTypes().size() > 1) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        private RowExpression reduce(FunctionMetadata aggregationFunctionMetadata, List<RowExpression> arguments)
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java:245]: `evaluateMinMax`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java [hunk 5]
```diff
@@ -151,31 +166,111 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> cont
...
+                            .map(TypeSignatureProvider::new)
+                            .collect(toImmutableList());
+                    Object reducedValue = evaluateConstantRowExpression(
+                            new CallExpression(
+                                    scalarFunctionName.getFunctionName(),
+                                    metadata.getFunctionManager().resolveFunction(Optional.empty(), scalarFunctionName, typeSignatureProviders),
+                                    returnType,
+                                    partitionedArguments),
+                            metadata,
+                            session.toConnectorSession());
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java:265]: Use the util in `Expressions` to build `CallExpression`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java [hunk 6]
```diff
@@ -151,31 +166,111 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> cont
...
+                    reducedArguments.add(constant(reducedValue, returnType));
+                }
+                arguments = reducedArguments;
+            }
+            if (arguments.isEmpty()) {
+                return constant(null, returnType);
+            }
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java:272]: When will the result be null?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java (hunk 2 above)
Comment (shixuan-fan) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java:169]: I extracted the newly added code into `reduce` (after changing `reduce` to `evaluateMinMax`). Also flipped the if statement so it looks like early return pattern.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java (hunk 3 above)
Comment (shixuan-fan) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java:203]: This is following the existing code, but I agree with the proposed change. Also changed the original code as well.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java (hunk 5 above)
Comment (shixuan-fan) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java:265]: Not aware of this. Looks so much better.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java (hunk 6 above)
Comment (shixuan-fan) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java:272]: The result would be null if all values are null.

Comment (shixuan-fan) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java:280]: That being said, I should probably move this to be the first step of this function.

**Pull request merged**